mavros/src/plugins/imu.cpp

```diff
@@ -0,0 +1,530 @@
+	{
+		PluginBase::initialize(uas_);
+
+		double linear_stdev, angular_stdev, orientation_stdev, mag_stdev;
+
+		/**
+		 * @warning A rotation from the aircraft-frame to the base_link frame is applied.
+		 * Additionally, it is reported the orientation of the vehicle to describe the
+		 * transformation from the ENU frame to the base_link frame (ENU <-> base_link).
+		 * THIS ORIENTATION IS NOT THE SAME AS THAT REPORTED BY THE FCU (NED <-> aircraft).
+		 */
+		imu_nh.param<std::string>("frame_id", frame_id, "base_link");
+		imu_nh.param("linear_acceleration_stdev", linear_stdev, 0.0003);		// check default by MPU6000 spec
+		imu_nh.param("angular_velocity_stdev", angular_stdev, 0.02 * (M_PI / 180.0));	// check default by MPU6000 spec
+		imu_nh.param("orientation_stdev", orientation_stdev, 1.0);
+		imu_nh.param("magnetic_stdev", mag_stdev, 0.0);
+
+		setup_covariance(linear_acceleration_cov, linear_stdev);
+		setup_covariance(angular_velocity_cov, angular_stdev);
+		setup_covariance(orientation_cov, orientation_stdev);
+		setup_covariance(magnetic_cov, mag_stdev);
+		setup_covariance(unk_orientation_cov, 0.0);
+
+		imu_pub = imu_nh.advertise<sensor_msgs::Imu>("data", 10);
+		magn_pub = imu_nh.advertise<sensor_msgs::MagneticField>("mag", 10);
+		temp_pub = imu_nh.advertise<sensor_msgs::Temperature>("temperature", 10);
+		press_pub = imu_nh.advertise<sensor_msgs::FluidPressure>("atm_pressure", 10);
+		imu_raw_pub = imu_nh.advertise<sensor_msgs::Imu>("data_raw", 10);
+
+		// Reset has_* flags on connection change
+		enable_connection_cb();
+	}
+
+	Subscriptions get_subscriptions() {
+		return {
+			       make_handler(&IMUPlugin::handle_attitude),
+			       make_handler(&IMUPlugin::handle_attitude_quaternion),
+			       make_handler(&IMUPlugin::handle_highres_imu),
+			       make_handler(&IMUPlugin::handle_raw_imu),
+			       make_handler(&IMUPlugin::handle_scaled_imu),
+			       make_handler(&IMUPlugin::handle_scaled_pressure),
+		};
+	}
+
+private:
+	ros::NodeHandle imu_nh;
+	std::string frame_id;
+
+	ros::Publisher imu_pub;
+	ros::Publisher imu_raw_pub;
+	ros::Publisher magn_pub;
+	ros::Publisher temp_pub;
+	ros::Publisher press_pub;
+
+	bool has_hr_imu;
+	bool has_scaled_imu;
+	bool has_att_quat;
+	Eigen::Vector3d linear_accel_vec_enu;
+	Eigen::Vector3d linear_accel_vec_ned;
+	ftf::Covariance3d linear_acceleration_cov;
+	ftf::Covariance3d angular_velocity_cov;
+	ftf::Covariance3d orientation_cov;
+	ftf::Covariance3d unk_orientation_cov;
+	ftf::Covariance3d magnetic_cov;
+
+	/* -*- helpers -*- */
+
+	/**
+	 * @brief Setup 3x3 covariance matrix
+	 * @param[in,out] cov	Covariance matrix
+	 * @param[in] stdev	Standard deviation
+	 * @remarks		Diagonal computed from the stdev
+	 */
+	void setup_covariance(ftf::Covariance3d &cov, double stdev)
+	{
+		ftf::EigenMapCovariance3d c(cov.data());
+
+		c.setZero();
+		if (stdev) {
+		    double sr = stdev * stdev;
+		    c.diagonal() << sr, sr, sr;
+		}
+		else {
+		   c(0,0) = -1.0;
+		}
```

Comment:
Spaces instead of tabs. Check uncrustify.

Reply:
sure